src/averageAges.js

```diff
@@ -1,61 +1,78 @@
 'use strict';
 
+function getAverage(people) {
+  const sumAges = people.reduce((sum, pers) => {
+    const persAge = pers.died - pers.born;
+
+    return sum + persAge;
+  }, 0);
+
+  return (sumAges / people.length);
+}
+
 /**
- * Implement calculateMenAverageAge function
- *
- * Function returns average age of men in array. If `century` is specified then
- * function calculates average age only for men who died in this century
- *
- * To calculate century:
- * Divide year of person's death by 100: Math.ceil(person.died / 100)
- *
  * @param {object[]} people
  * @param {number} century - optional
  *
  * @return {number}
  */
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const men = people.filter(
+    century
+      ? pers => {
+        const validCentury = Math.ceil(pers.died / 100) === century;
+
+        return validCentury && pers.sex === 'm';
+      }
+      : pers => pers.sex === 'm');
+
+  return getAverage(men);
 }
 
 /**
- * Implement calculateWomenAverageAge function
- *
- * Function returns average age of women in array. If `withChildren` is
- * specified then function calculates average age only for women with children
- *
- * Hint: To check if a woman has children you should find someone who mention
- * her as mother.
- *
  * @param {object[]} people
  * @param {boolean} withChildren - optional
  *
  * @return {number}
  */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const women = people.filter(withChildren
+    ? pers => {
+      const validChild = people.find(child =>
+        child.mother === pers.name);
+
+      return validChild && pers.sex === 'f';
+    }
+    : pers => pers.sex === 'f');
+
+  return getAverage(women);
 }
 
 /**
- * Implement calculateAverageAgeDiff function.
- *
- * The function returns an average age difference between a child and his or her
- * mother in the array. (A mother's age at child birth)
- *
- * If `onlyWithSon` is specified then function calculates age difference only
- * for sons and their mothers.
- *
  * @param {object[]} people
  * @param {boolean} onlyWithSon - optional
  *
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const validMother = (child) => {
+    const findMother = (people.find(mother =>
+      child.mother === mother.name));
+
+    return onlyWithSon
+      ? findMother && child.sex === 'm'
+      : findMother;
+  };
+
+  const childsWithMother = people.filter(validMother);
+  const sumOfAges = childsWithMother.reduce((sum, pers) => {
+    const diffAge = pers.born - people.find(mother =>
+      pers.mother === mother.name).born;
```

Comment:
not optimal, you've already been looking for mother at line 59-60
if you have found mothers, then it makes sense to save them, and not look for them again